vortex-duckdb/src/exporter/list.rs

```diff
@@ -0,0 +1,183 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright the Vortex contributors
+
+use std::marker::PhantomData;
+
+use itertools::Itertools as _;
+use vortex::ToCanonical as _;
+use vortex::arrays::{ListArray, PrimitiveArray};
+use vortex::dtype::{NativePType, match_each_integer_ptype};
+use vortex::error::{VortexResult, vortex_err};
+use vortex::mask::Mask;
+
+use crate::cpp;
+use crate::duckdb::Vector;
+use crate::exporter::{ColumnExporter, VectorExt};
+
+use super::{ConversionCache, new_array_exporter};
+
+struct ListExporter<T> {
+    validity: Mask,
+    elements_exporter: Box<dyn ColumnExporter>,
+    offsets: PrimitiveArray,
+    offset_type: PhantomData<T>,
+}
+
+pub(crate) fn new_exporter(
+    array: &ListArray,
+    cache: &ConversionCache,
+) -> VortexResult<Box<dyn ColumnExporter>> {
+    let elements_exporter = new_array_exporter(array.elements(), cache)?;
+    let offsets = array.offsets().to_primitive()?;
+    let boxed = match_each_integer_ptype!(offsets.ptype(), |T| {
+        Box::new(ListExporter {
+            validity: array.validity_mask()?,
+            elements_exporter,
+            offsets,
+            offset_type: PhantomData::<T>,
+        }) as Box<dyn ColumnExporter>
+    });
+    Ok(boxed)
+}
+
+impl<T: NativePType> ColumnExporter for ListExporter<T> {
+    fn export(&self, offset: usize, len: usize, vector: &mut Vector) -> VortexResult<()> {
+        // Set validity if necessary.
+        if unsafe { vector.set_validity(&self.validity, offset, len) } {
+            // All values are null, so no point copying the data.
+            return Ok(());
+        }
+
+        let offsets = &self.offsets.as_slice::<T>()[offset..][..(len + 1)];
+        let start_offset = offsets[0]
+            .to_u64()
+            .ok_or_else(|| vortex_err!("list offsets must fit in u64"))?;
+        let mut sum_of_list_lens = 0_u64;
+
+        let offsets_slice: &mut [cpp::duckdb_list_entry] =
+            unsafe { vector.as_slice_mut::<cpp::duckdb_list_entry>(len) };
+
+        for (window, destination) in offsets.windows(2).zip_eq(offsets_slice) {
+            let start = window[0]
+                .to_u64()
+                .ok_or_else(|| vortex_err!("list offsets must fit in u64"))?;
+            let end = window[1]
+                .to_u64()
+                .ok_or_else(|| vortex_err!("list offsets must fit in u64"))?;
+            let len = end - start;
+            sum_of_list_lens += len;
+            *destination = cpp::duckdb_list_entry {
+                offset: start - start_offset,
+                length: len,
+            };
+        }
+
+        vector.list_vector_reserve(sum_of_list_lens)?;
```

Comment:
This is not a blocker, but the multiple calls to reserve (one for each export) seems unnecessarily you could allocate once a larger vector and export the values once

Reply:
I added a TODO